interfaces/prompting/patterns/render.go

```diff
@@ -0,0 +1,376 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package patterns
+
+import (
+	"bytes"
+	"strings"
+)
+
+// RenderConfig is a configuration of a render node.
+type RenderConfig interface {
+	// NextEx modifies the configuration to the next state, if any remain.
+	// Returns the length in bytes of the new configuration, the number of
+	// bytes which remain unchanged since the previous configuration, and true
+	// if more configurations remain to be rendered. The argument is always the
+	// render node that was earlier used to obtain render configuration.
+	NextEx(n RenderNode) (length, lengthUnchanged int, moreRemain bool)
+	// Length returns the total length of the rendered node in its current
+	// configuration, including all sub-nodes if the node is a seq or alt.
+	// The argument is always the render node that was earlier used to obtain
+	// the render configuration.
+	Length(n RenderNode) int
+}
+
+// RenderNode is a node which may be rendered in a particular configuration.
+type RenderNode interface {
+	// Render renders the given configuration to the buffer if alreadyWritten
+	// equals 0. Otherwise, subtracts from alreadyWritten the length of the
+	// string which would have been written to the buffer, and returns the
+	// difference.
+	Render(buf *bytes.Buffer, conf RenderConfig, alreadyWritten int) int
+	// Config returns the initial render configuration.
+	Config() RenderConfig
+	// NumVariants returns the number of variants this node can be rendered as (recursively).
+	NumVariants() int
+	// nodeEqual returns true if two nodes are recursively equal.
+	nodeEqual(RenderNode) bool
+}
+
+// RenderAllVariants renders subsequent variants to a reusable buffer.
```

Comment:
this needs to explain how observe is exactly called for and what we mean by Variant. From an external POV this comment also apply to the actualy method on PathPattern that uses this